### PR TITLE
tensorflow plugin implementation

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
TensorFlow plugin for Flyte that leverages https://github.com/kubeflow/tf-operator to run distributed tensorflow jobs on k8s.

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [x] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
related to https://github.com/lyft/flyteidl/pull/71

## Tracking Issue
https://github.com/lyft/flyte/issues/115

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by pingsutw** [comment]
The test failure seems not related to PR.

**Comment by kumare3** [comment]
I will take a look at the test

**Comment by pingsutw** [comment]
Thanks :)

File: go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow_test.go
Comment (swiftdiaries) [inline go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow_test.go:73]: s/dummySparkTaskTemplate/dummyTensorflowTaskTemplate?

File: go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow_test.go
Comment (swiftdiaries) [inline go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow_test.go:276]: Same as the comment above

**Comment by swiftdiaries** [comment]
Otherwise, LGTM :)

**Comment by pingsutw** [comment]
@swiftdiaries sorry for the late reply.
I've updated and rebased the PR

**Comment by swiftdiaries** [comment]
LGTM 
I don't have write access to this repo, I think @kumare3 might want to do a round of reviews before merging it

**Comment by swiftdiaries** [comment]
This is really cool 😬 Great to see this come along

**Comment by kumare3** [comment]
Hey I thought this was merged. One check seems
To be failing?

**Comment by pingsutw** [comment]
Test failure seems unrelated. 
btw, I'm going to write an example in flytekit.

**Comment by kumare3** [comment]
Cc @wild-endeavor @EngHabu @katrogan

File: go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go
Comment (kumare3) [reply go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go:24]: @pingsutw this is not needed here right?

File: go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go
Comment (kumare3) [inline go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go:144]: can we share this code between all kf operators like - 
pytorch and tensorflow. they seem to be the same?

**Comment by kumare3** [comment]
hmm ya the test is not related, it is interesting to me that your branch has this error, but master does not?

File: go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go
Comment (pingsutw) [reply go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go]: Tensorflow operator will also force pod to have a container named `tensorflow`, I've updated the comment.

**Comment by pingsutw** [comment]
I've moved some common code between PyTorch and TensorFlow to `common_opeartor`

**Comment by pingsutw** [comment]
> hmm ya the test is not related, it is interesting to me that your branch has this error, but master does not?

I got the same error in master
https://github.com/pingsutw/flyteplugins/actions/runs/215382594

**Comment by kumare3** [comment]
Let me review today and we can merge soon

**Review by kumare3 — APPROVED** [review]
LGTM

**Comment by kumare3** [comment]
Some files are not go-imported. Once they are done, the linter should be green and let us merge it afterwards. I do not have permissions to your fork.

**Comment by pingsutw** [comment]
cc @EngHabu @swiftdiaries @kumare3 Can you help me review this, Thanks.
I've fixed the conflict and the PR tests and lint already passed.

However, I don't why `Build Co-Pilot Docker image` failed.
Do you know how to fix it?
```
Removing intermediate container b9a18bc4bc63
 ---> 943a353f299a
Successfully built 943a353f299a
Successfully tagged my_awesome_image:latest
+ set +x

[Action Step] Tagging image...
Tag: ghcr.io/flyteorg/flytecopilot:1a52a8f80c80711ab2175ce3b58db715ad13af59

[Action Step] Pushing image...
ERROR: Can't push when not logged in to registry. Set "push_image_and_stages: false" if you don't want to push
```

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (EngHabu) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:23]: Does this serve any function?
```suggestion
	sort.Slice(jobConditions, func(i, j int) bool {
```

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (EngHabu) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:24]: nil checks?

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (EngHabu) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:48]: Restarting means it's already going to get retried on its own, right? returning RetryableFailure would actually make flyte kill the existing instance and retry...
If my understanding is correct, this should return Running

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (EngHabu) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:58]: You can use similar code to this: https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/logs/logging_utils.go#L45-L53
to avoid checking log plugins yourself... as long as you have the names of the pods...

File: go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go
Comment (EngHabu) [inline go/tasks/plugins/k8s/kfoperators/tensorflow/tensorflow.go:139]: Should this use `common.GetPhaseInfo` you defined above?

**Review by EngHabu — COMMENTED** [review]
Awesome work, @pingsutw. Just have a few comments and we can merge!

**Comment by pingsutw** [comment]
@EngHabu Thanks for your review. I will update ASAP.

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (pingsutw) [reply go/tasks/plugins/k8s/kfoperators/common/common_operator.go]: Done

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (pingsutw) [reply go/tasks/plugins/k8s/kfoperators/common/common_operator.go:51]: This snippet of code is copied from https://github.com/flyteorg/flyteplugins/blob/8c66bddd184ea1d605e74ea79177df7a4fdf124c/go/tasks/plugins/k8s/kfoperators/pytorch/pytorch.go#L134-L149

I think if the job was failed maybe there is something wrong like the docker image can not pull, so we should add a job failed message to the log.
btw, if I change to `pluginsCore.PhaseInfoRunnin`, the test will be broken.
If we make sure to change to `running`, I will change test sets as well.

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (pingsutw) [reply go/tasks/plugins/k8s/kfoperators/common/common_operator.go]: Done

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (pingsutw) [reply go/tasks/plugins/k8s/kfoperators/common/common_operator.go]: Done

**Comment by pingsutw** [comment]
@EngHabu and @kumare3 I've updated it. I think it's ready to merge.

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (kumare3) [reply go/tasks/plugins/k8s/kfoperators/common/common_operator.go:51]: @pingsutw do you know of cases in which Operator status will be restarting? I can maybe look at the code. 
But @EngHabu I think I agree that Pytorch and TFoperator should be the same as Kf has common status across both.

**Comment by kumare3** [comment]
I think I am +1 on the issue, will wait for @EngHabu to approve.
On the other hand @pingsutw we also need to add an example here - https://github.com/flyteorg/flytesnacks/tree/push-images/cookbook/plugins

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (EngHabu) [reply go/tasks/plugins/k8s/kfoperators/common/common_operator.go:51]: I understand @pingsutw this is copied code. I'm not sure this particular case was tested before to be honest.
[This is the snippet ](https://github.com/kubeflow/tf-operator/blob/4f6906951804ff4530ffe2cb9567b19f1eeb760d/pkg/controller.v1/tensorflow/pod.go#L135-L157) of the code that sets the status to restarting.
It sounds like from Flyte perspective, it'll keep running... 

So yes, we will also need to update the test.

Are you referring to [this](https://github.com/kubeflow/common/blob/master/pkg/apis/common/v1/types.go#L118-L122) for the common code?

I agree we should replicate what kf operators common pattern has...

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (EngHabu) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:74]: I missed this in my previous review... I don't think you need this line any more... just call GetTaskLogs (like you do below) and pass as much information as you can... let the log plugin take care of which log links to produce...

Comment (pingsutw) [reply go/tasks/plugins/k8s/kfoperators/common/common_operator.go]: Sorry, I forgot to remove it. And I just updated it.

**Comment by EngHabu** [comment]
Let's address this: https://github.com/flyteorg/flyteplugins/pull/103#discussion_r575277026 then let's get it merged!

**Comment by pingsutw** [comment]
> Let's address this: [#103 (comment)](https://github.com/flyteorg/flyteplugins/pull/103#discussion_r575277026) then let's get it merged!

okay, I'm working on it.

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go
Comment (pingsutw) [reply go/tasks/plugins/k8s/kfoperators/common/common_operator.go:51]: I've updated to `pluginsCore.PhaseInfoRunnin`

**Review by EngHabu — APPROVED** [review]

**Pull request merged**
